Initial checkin of RhythmOne bidder adapter and unit tests

jstocker76 opened the pull request:
<!--
Thank you for your pull request. Please make sure this PR is scoped to one change, and that any added or changed code includes tests with greater than 80% code coverage. See http://prebid.org/dev-docs/testing-prebid.html for documentation on testing Prebid.js.
-->
## Type of change

<!-- Remove items that don't apply and/or select an item by changing [ ] to [x] -->
- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other
## Description of change

Addition of a new bidder adapter that uses RhythmOne's ad exchange with unit tests.

<!-- For new bidder adapters, please provide the following -->
- test parameters for validating bids

```
			{
				bidder: 'rhythmone',
				params: {
					placementId: "549",
					zone: "hb",
					path: "mvo"
				}
			}
```
- contact email of the adapter’s maintainer
  astocker@rhythmone.com
- [x] official adapter submission

KevinCHill commented:
Hi, Thank you for contacting me. I'm out of the office for the rest of the  
day. Though I do check my emails randomly throughout the evening, I will  
most likely get back to you in the morning

jstocker76 commented:
oy.  my bad.  Will fix then submit another pull

jstocker76 commented:
My unit tests were validating a bid property that wasn't under my direct control (bid.type) and had been removed since my last pull from master.  I removed this.  Should be good to go

ojotoxy commented:
Not sure if it's appropriate to explicitly refer to "pbjs" in an adapter anymore, we switched to using "$$PREBID_GLOBAL$$" which may be configured when building the project.

jstocker76 commented:
Aah, got it.  I've replaced uses of "pbjs" with the $$PREBID_GLOBAL$$ macro.

protonate commented:
@jstocker76 thanks for the PR. I'm not getting the ad to render with the provided test parameters. I am getting bids and ads back. I'm using `localhost` and `prebidapp.com` domains for testing. Any ideas?

Also see #509 about matching bid request IDs with bid response IDs.

mkendall07 commented on src/adapters/rhythmone.js:
```diff
@@ -0,0 +1,496 @@
+  
+  var rhythmBidderUtilities = {
+    template: function(){return "<div id=\"{0}_wrapper\"></div><script type=\"text/javascript\">("+rhythmBidderUtilities.process.toString()+")({1}, {6}, {2}, {3}, \"{0}\", {4}, \"{5}\");</scr"+"ipt>";},
+    process: "function (e,t,r,n,i,a,o){function d(e,t,r,i){var a=[\"bidderCode\",\"size\",\"adUnitCode\",\"cpm\",\"timeToRespond\"],o=[\"b\",\"s\",\"a\",\"c\",\"t\"],d=0,s=[],c=new Image,u=\"\",p=document.location.ancestorOrigins,l=\"//rhythmanalytics.1rx.io/rhythmanalytics/img?tracking_id=prebid_r1&hit_type=event&time_ms=\"+(new Date).getTime()+\"&metric0=allBids&metric1=\"+encodeURIComponent(t) ...
```
Why is the code minified here?

mkendall07 commented on src/adapters/rhythmone.js:
```diff
@@ -0,0 +1,496 @@
+    
+    logToConsole("posting "+json+" to "+url);
+    
+    if(typeof global.XDomainRequest !== "undefined"){
```
We have utility for making ajax requests - please use `https://github.com/prebid/Prebid.js/blob/master/src/ajax.js` instead

mkendall07 commented on src/adapters/rhythmone.js:
```diff
@@ -0,0 +1,496 @@
+
+  var bidderCode = "rhythmone";
+  
+  function GUID() {
```
pleas use `utils.generateUUID()` instead.

mkendall07 commented on src/adapters/rhythmone.js:
```diff
@@ -0,0 +1,496 @@
+    setIfPresent(o.site, "page", function(){
+      var d = (typeof bid.params === "object" ? bid.params.domain : null),
+        l;
+      try{l = top.document.location.href.toString();}
```
please use `window.top`

mkendall07 commented on src/adapters/rhythmone.js:
```diff
@@ -0,0 +1,496 @@
+      d = document.location.ancestorOrigins;
+      if(d && d.length > 0)
+        return d[d.length-1];
+      return top.document.location.hostname;
```
Please use `window.top` and add try/catch.

mkendall07 commented on src/adapters/rhythmone.js:
```diff
@@ -0,0 +1,496 @@
+        o.site.keywords = bid.params.keywords;
+    }
+    
+    o.device.devicetype = ((/(ios|ipod|ipad|iphone|android)/i).test(global.navigator.userAgent) ? 1 : ((/(smart[-]?tv|hbbtv|appletv|googletv|hdmi|netcast\.tv|viera|nettv|roku|\bdtv\b|sonydtv|inettvbrowser|\btv\b)/i).test(global.navigator.userAgent) ? 3 : 2));
```
Consider adding this as a function in `utils` as it might be useful for others.

mkendall07 commented on src/adapters/rhythmone.js:
```diff
@@ -0,0 +1,496 @@
+  
+    function finish(){
+  
+      var sandbox = new ZTStorage("//d3rim8qxq4v92b.cloudfront.net/ztstorage/bg.htm"),
```
can you explain what this is used for?

mkendall07 commented on adapters.json:
```diff
@@ -14,6 +14,7 @@
     "piximedia",
     "pubmatic",
     "pulsepoint",
+	"rhythmone",
```
please fix spacing

jstocker76 commented:
There have been some unforseen alterations to the ad api that require a lot of changes (mostly reductions) in code for this bidder adapter.  I'm going to close this pull request until the time comes when the API is stable, and resubmit with all your feedback that's still relevant.  Thanks very much for the code reviews

jstocker76 commented:
The bidder object has been updated with new test parameters-

{
	bidder: 'rhythmone',
	params: {
		placementId: "549",
		zone: "hb",
		path: "mvo"
	}
}

Alterations since the last pull request:

1) The ad API can now handle multiple impressions.  I was previously issuing a separate ad request for every combination of ad slot and size due to the limitations of the API.  These  have been consolidated into 1 request

2) Removed dependencies on local browser storage.  During previous code reviews there were questions about what the file "zstorage.js" did.  It was used to store a persistent user id (and a couple other things) in local browser storage via a sandboxed iframe.  This is no longer necessary

3) Added support for a GET request, avoiding the separate OPTIONS request.

4) Removed the need to ping an API to retrieve the user's IP


Code review/feedback addressed:

mkendall07 - "Why is the code minified here?"
The ad markup is wrapped in a script that applies ORTB macros when the ad is rendered.  I had kept this as a template string because none of this code is executed by the bidder adapter directly.

mkendall07 - "We have utility for making ajax requests - please use https://github.com/prebid/Prebid.js/blob/master/src/ajax.js instead"
done.

mkendall07 - "pleas use utils.generateUUID() instead."
done.

mkendall07 - "please use window.top"
done.

mkendall07 - "Please use window.top and add try/catch."
Added window.top.  The try/catch is in place in the setIfPresent function.  It wraps execution of the callback method.

mkendall07 - "can you explain what this is used for?" (in reference to the zstorage.js file)
dependencies on this file have since been removed.  It was a utility that streamlines saving data to local browser storage on a common domain.  I was using it to store a semi-permanent user id.

mkendall07 - "please fix spacing" (in adapters.json)
done.

I'm confident there won't be any major changes coming from my end from this point onward.  Sorry for the delay.

mkendall07 commented on src/adapters/rhythmone.js:
```diff
@@ -0,0 +1,453 @@
+          catch(ex){}
+          return null;
+        })(),
+        responses = (pjs && pjs.getBidResponses ? pjs.getBidResponses() : {}),
```
adapters should not access other bid responses. Can you please remove this bit

jstocker76 replied:
That makes it rather difficult to populate the {AUCTION_PRICE} ortb macro...

jstocker76 replied:
I've talked this over with the powers that be, and will remove the ortb macro substitution.

abkbynature replied:
Is there a policy or guideline that states what public objects/data should not be accessed?

mkendall07 replied:
No, but we should have one. Essentially no adapter should be looking at bids submitted by other partners or otherwise using the public API's designated for publishers. What this means practically is that you do not need to access any methods/properties on `pbjs` with the exception of if your adapter needs to expose a JSONP callback handler function.

abkbynature replied:
What about registering to pbjs notifications such as auctionEnd, timeout, etc?

jstocker76 replied:
Done.

mkendall07 replied:
@abkbynature 
I'll review the code here and let you know if I have any objections.

protonate commented on src/adapters/rhythmone.js:
```diff
@@ -0,0 +1,360 @@
+/*
+  Copyright (c) 2016 RhythmOne, LLC. All rights reserved.
```
Please remove, all submissions are subject to the Apache 2.0 License.
https://github.com/prebid/Prebid.js/blob/master/LICENSE

protonate commented on src/adapters/rhythmone.js:
```diff
@@ -0,0 +1,360 @@
+
+import {ajax as ajax} from '../ajax';
+
+//function setupGA() {
```
Please remove commented code.

protonate commented on src/adapters/rhythmone.js:
```diff
@@ -0,0 +1,360 @@
+  function noBids(params){
+    for(var i=0; i<params.bids.length; i++){
+      if(params.bids[i].success !== 1){
+        var bid = bidfactory.createBid(2);
```
Adapters can now pass a constant for status and the originating bid request, the first for readability and the second to match bid request / response IDs per #509.

e.g. 
```
var bid = bidfactory.createBid(CONSTANTS.STATUS.NO_BID, bidRequest);
```

protonate commented on src/adapters/rhythmone.js:
```diff
@@ -0,0 +1,360 @@
+            
+              slotMap[bid.impid].success = 1;
+              
+              var pbResponse = bidfactory.createBid(1),
```
Adapters can now pass a constant for status and the originating bid request, the first for readability and the second to match bid request / response IDs per #509.

e.g. 
```
var bid = bidfactory.createBid(CONSTANTS.STATUS.GOOD, bidRequest);
```

protonate commented:
Thanks for the updates, reviewed and approved.

aneuway2 commented:
@jstocker76 are all parameter's required or are there optional inputs?

jstocker76 commented:
aneuway2 - only placementId is required

mkendall07 commented:
@jstocker76 
Please submit a docs PR here: https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders

jstocker76 commented:
mkendall07 - done

aneuway2 commented:
@jstocker76 looking at http://prebid.org/dev-docs/bidders.html#rhythmone and the PR for documentation it looks like *placementId* was specified in the documentation instead of *publisherId*. is this an error, or did it change to placementId?

Thanks!

jstocker76 commented:
"placementId" is what the code expects.  Whether, in practice, it functions as a placement id or publisher id depends on how the publisher's account is set up in our system.
